src/cloud-api-adaptor/pkg/adaptor/cloud/cloud.go

```diff
@@ -286,7 +288,46 @@ func (s *cloudService) CreateVM(ctx context.Context, req *pb.CreateVMRequest) (r
 		},
 	}
 
-	if s.aaKBCParams != "" {
+	if authJSON != nil {
+		if len(authJSON) > cloudinit.DefaultAuthfileLimit {
+			logger.Printf("Credentials file is too large to be included in cloud-config")
+		} else {
+			cloudConfig.WriteFiles = append(cloudConfig.WriteFiles, cloudinit.WriteFile{
+				Path:    AuthFilePath,
+				Content: string(authJSON),
+			})
+		}
+	}
+
+	initdataStr := util.GetInitdataFromAnnotation(req.Annotations)
+	logger.Printf("initdata: %s", initdataStr)
+	if initdataStr != "" {
+		decodedBytes, err := base64.StdEncoding.DecodeString(initdataStr)
+		if err != nil {
+			return nil, fmt.Errorf("Error base64 decode initdata: %w", err)
+		}
+		initdata := userdata.InitData{}
+		err = toml.Unmarshal(decodedBytes, &initdata)
+		if err != nil {
+			return nil, fmt.Errorf("Error unmarshalling initdata: %w", err)
+		}
+		for key, value := range initdata.Data {
+			cloudConfig.WriteFiles = append(cloudConfig.WriteFiles, cloudinit.WriteFile{
+				Path:    filepath.Join(userdata.ConfigParent, key),
+				Content: value,
+			})
+		}
+		cloned := initdata
+		cloned.Data = nil
+		clonedToml, err := toml.Marshal(cloned) // Data field omitted
+		if err != nil {
+			return nil, fmt.Errorf("Error Marshal cloned initdata: %w", err)
+		}
+		cloudConfig.WriteFiles = append(cloudConfig.WriteFiles, cloudinit.WriteFile{
+			Path:    userdata.InitdataMeta,
+			Content: string(clonedToml),
+		})
+	} else if s.aaKBCParams != "" { // Keep AA_KBC_PARAMS support as it is used by e2e test, KBS is dynamic k8s service in e2e test
```

Comment:
It's fine to keep the AA_KBC_PARAMS in this PR, but we should have follow up TODO story to fully remove AA_KBC_PARAMS support with e2e-test cases update.
maybe update the comment add a TODO/FIXME keyword?